core/drivers/crypto/hisilicon/sec_main.c

```diff
@@ -0,0 +1,274 @@
+		if (free_qp_num > max_qp_num) {
+			max_qp_num = free_qp_num;
+			sec_dev = cur_dev;
+		}
+	}
+
+	if (!sec_dev) {
+		EMSG("No available sec device");
+		return NULL;
+	}
+
+	return hisi_qm_create_qp(&sec_dev->qm, sq_type);
+}
+
+static void sec_enable_clock_gate(struct hisi_qm *qm)
+{
+	if (qm->version == HISI_QM_HW_V2)
+		return;
+
+	io_setbits32(qm->io_base + SEC_CONTROL_REG, SEC_CLK_GATE_ENABLE);
+	io_write32(qm->io_base + SEC_DYNAMIC_GATE_V3, SEC_DYNAMIC_GATE_EN);
+	io_write32(qm->io_base + SEC_CORE_AUTO_GATE_V3, SEC_CORE_AUTO_GATE_EN);
+}
+
+static enum hisi_drv_status sec_engine_init(struct acc_device *sec_dev)
+{
+	struct hisi_qm *qm = &sec_dev->qm;
+	uint32_t val = 0;
+
+	if (qm->fun_type == HISI_QM_HW_VF)
+		return HISI_QM_DRVCRYPT_NO_ERR;
+
+	/* QM_HW_V2 version need to close clock gating */
+	io_clrbits32(qm->io_base + SEC_CONTROL_REG, SEC_CLK_GATE_ENABLE);
+
+	hisi_qm_dev_init(qm);
+
+	io_write32(qm->io_base + SEC_MEM_START_INIT, 0x1);
+	if (IO_READ32_POLL_TIMEOUT(qm->io_base + SEC_MEM_INIT_DONE, val,
+				   val & 0x1, POLL_PERIOD, POLL_TIMEOUT)) {
+		EMSG("Fail to init sec mem");
+		return HISI_QM_DRVCRYPT_ETMOUT;
+	}
+
+	io_setbits32(qm->io_base + SEC_CONTROL_REG, sec_dev->endian);
+
+	if (qm->version == HISI_QM_HW_V2) {
+		/* SMMU bypass */
+		io_write32(qm->io_base + SEC_INTERFACE_USER_CTRL0,
+			   SEC_USER0_CFG);
+		io_write32(qm->io_base + SEC_INTERFACE_USER_CTRL1,
+			   SEC_USER1_CFG);
+		io_write32(qm->io_base + AM_CFG_SINGLE_PORT_MAX_TRANS,
+			   SEC_SINGLE_PORT_MAX_TRANS);
+		io_write32(qm->io_base + SEC_SAA_EN, SEC_SAA_ENABLE);
+		/* HW V2 enable SM4 extra mode, as CTR/ECB */
+		io_write32(qm->io_base + SEC_BD_ERR_CHK_EN0, SEC_BD_ERR_CHK0);
+		/* Enable SM4 xts mode multiple iv */
+		io_write32(qm->io_base + SEC_BD_ERR_CHK_EN1, SEC_BD_ERR_CHK1);
+		/* disable PBKDF2 len check */
+		io_write32(qm->io_base + SEC_BD_ERR_CHK_EN2, SEC_BD_ERR_CHK2);
+		io_write32(qm->io_base + SEC_BD_ERR_CHK_EN3, SEC_BD_ERR_CHK3);
+	} else {
+		/* cmd_type is controlled by HAC subctrl, default normal */
+		io_write32(qm->io_base + SEC_INTERFACE_USER_CTRL0_V3,
+			   SEC_USER0_CFG_V3);
+		io_write32(qm->io_base + SEC_INTERFACE_USER_CTRL1_V3,
+			   SEC_USER1_CFG_V3);
+		io_write32(qm->io_base + SEC_SAA_EN, SEC_SAA_ENABLE_V3);
+		/* disable PBKDF2 salt len check */
+		io_write32(qm->io_base + SEC_BD_ERR_CHK_EN2_V3,
+			   SEC_BD_ERR_CHK2);
+	}
+	io_write32(qm->io_base + SEC_RAS_CE_ENABLE, SEC_RAS_CE_ENB_MASK);
+	io_write32(qm->io_base + SEC_RAS_FE_ENABLE, SEC_RAS_FE_ENB);
+	io_write32(qm->io_base + SEC_RAS_NFE_ENABLE, SEC_RAS_NFE_ENB_MASK);
+	io_write32(qm->io_base + SEC_CORE_INT_MASK, SEC_ABNML_INT_DISABLE);
+
+	sec_enable_clock_gate(qm);
+
+	return HISI_QM_DRVCRYPT_NO_ERR;
+}
+
+static enum hisi_drv_status sec_dev_status_check(struct hisi_qm *qm)
+{
+	uint32_t val = 0;
+
+	val = io_read32(qm->io_base + SEC_CORE_INT_SOURCE);
+	if (val & SEC_RAS_NFE_ENB_MASK) {
+		EMSG("SEC NFE RAS happened, need to reset");
```

Comment:
Is someone supposed to read this and do something?

Reply:
Please use DMSG() for debug info.

Reply:
The function returns an error code so I think it's fair it uses here the `EMSG()` macro.

Reply:
I think so, and  I didn't make the change.